src/lib/io/include/sol-efivarfs-storage.h

```diff
@@ -36,114 +36,161 @@
 #include <stdint.h>
 
 #include "sol-buffer.h"
+#include "sol-log.h"
 #include "sol-types.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-int sol_efivars_write_raw(const char *name, const struct sol_buffer *buffer);
+int sol_efivars_write_raw(const char *name, struct sol_blob *blob,
+    void (*cb)(void *data, const char *name, struct sol_blob *blob, int status),
+    const void *data);
 int sol_efivars_read_raw(const char *name, struct sol_buffer *buffer);
 
-#define CREATE_BUFFER(_val, _empty) \
+#define CREATE_BUFFER(_val) \
     struct sol_buffer buf = SOL_BUFFER_INIT_FLAGS(_val, \
-    sizeof(*(_val)), SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED); \
-    buf.used = (_empty) ? 0 : sizeof(*(_val));
+    sizeof(*(_val)), SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED | SOL_BUFFER_FLAGS_NO_NUL_BYTE);
+
+#define CREATE_BLOB(_val) \
+    struct sol_blob *blob; \
+    size_t _s = sizeof(*_val); \
+    void *v = malloc(_s); \
+    SOL_NULL_CHECK(v, -EINVAL); \
+    memcpy(v, _val, _s); \
+    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, v, _s); \
+    SOL_NULL_CHECK(blob, -EINVAL);
```

Comment:
On Thu, 15 Oct 2015 13:41:18 -0700, Ederson de Souza wrote:
> >      struct sol_buffer buf = SOL_BUFFER_INIT_FLAGS(_val, \
> > -    sizeof(*(_val)), SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED); \
> > -    buf.used = (_empty) ? 0 : sizeof(*(_val));
> > +    sizeof(*(_val)), SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED | SOL_BUFFER_FLAGS_NO_NUL_BYTE);
> > +
> > +#define CREATE_BLOB(_val) \
> > +    struct sol_blob *blob; \
> > +    size_t _s = sizeof(*_val); \
> > +    void *v = malloc(_s); \
> > +    SOL_NULL_CHECK(v, -EINVAL); \
> > +    memcpy(v, _val, _s); \
> > +    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, v, _s); \
> > +    SOL_NULL_CHECK(blob, -EINVAL);
> 
> I don't think so. `sol_blob_new` takes care of it, when using `SOL_BLOB_TYPE_DEFAULT`.
> 

But if sol_blob_new() fails, it leaks.